### ISSUE TITLE
Using Activity for creating preference store in lazy could leak

### ISSUE THREAD

**Issue by kozaxinan** [ashdavies/preferences-store#3]
Store use activity to create SharedPreference with lazy lambda. It is an invitation to memory leaks. if Store is created with activity and hold in ViewModel, it will leak the activity.

Don't pass the activity to lazy lambda.

```
fun SharedPreferencesStore(
        activity: Activity
): SharedPreferencesStore ...
```

Here is leak trace of your code 😄 

```
a.b.c.finance.calculator.FViewModel instance
│    Leaking: UNKNOWN
│    ↓ FViewModel.cache
│                                 ~~~~~
├─ a.b.c.finance.calculator.FCache instance
│    Leaking: UNKNOWN
│    ↓ FCache.$$delegate_0
│                             ~~~~~~~~~~~~
├─ io.ashdavies.preferences.SharedPreferencesStoreKt$SharedPreferencesStore$2 instance
│    Leaking: UNKNOWN
│    Anonymous class implementing io.ashdavies.preferences.SharedPreferencesStore
│    ↓ SharedPreferencesStoreKt$SharedPreferencesStore$2.$activity
│                                                        ~~~~~~~~~
╰→ a.b.c.finance.FActivity instance
​     Leaking: YES (ObjectWatcher was watching this because a.b.c.finance.FActivity received Activity#onDestroy() callback and Activity#mDestroyed is true)
```

**Comment by ashdavies** [comment]
Using `Activity` is a perfectly acceptable method of creating `SharedPreferences`, the issue here is that the reference is leaked because of the lazy initialisation. The solution would be one of the following...
- Remove lazy initialisation of `SharedPreferences` (could introduce regressions)
- Introduce initialisation abstraction to allow user to opt-in or out (depends on ubiquity)
- Use `WeakReference` for `Activity` and throw if collected (seems lazy)

**Comment by kozaxinan** [comment]
```
@Suppress("FunctionName")
fun SharedPreferencesStore(
    activity: Activity
): SharedPreferencesStore = SharedPreferencesStore(activity.application, activity.localClassName)
```
changing method to this suggestion wont leak activity into lazy.

**Comment by ashdavies** [comment]
Agreed, that should work, thanks for the suggestion 👍

**Comment by ashdavies** [comment]
Should be resolved with 6633e1f66ec5879269aa4b3ff15a7999a4196d33
